In [1]:
## Project: Exploring Ebay Car Sales Data

In [2]:
# read file autos.csv

import pandas as pd

autos = pd.read_csv("autos.csv",encoding="Latin-1")


# 1) Task: Display information about dataframe autos

In [3]:
#display information about autos

print(autos.info())
print(autos.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

# 2) Task: Clean Column Names

In [4]:
#clean column names

print(autos.columns)


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
# assign cleaned columns names to a new array
# function created to clean columns

def clean_col(col):
    col = col.strip()
    col = col.replace("yearOfRegistration","registration_year")
    col = col.replace("monthOfRegistration", "registration_month")
    col = col.replace("notRepairedDamage", "unrepaired_damage")
    col = col.replace("dateCreated", "ad_created")
    col = col.lower()
    return col

new_columns = []
for c in autos.columns:
    clean_c = clean_col(c)
    new_columns.append(clean_c)
    
autos.columns = new_columns

#print new column names
print(autos.columns)



Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')


In [6]:
#check current state of the dataframe

print(autos.head())

           datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,00

### Note: The autos column have been cleaned and renamed. All column names converted from camelcase to snakecase.

# 3) Task: Initial Exploration and Cleaning

In [7]:
# describe autos dataframe

autos.describe(include='all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

## Notes for the cells in the following lines:
### - Any column that have mostly one value that are candidates to be dropped
### - Any column that need more investigation
### - Any example of numeric data stored as text  that needs to be cleaned 

In [8]:
# check for null columns

print(autos.isnull().sum())


datecrawled              0
name                     0
seller                   0
offertype                0
price                    0
abtest                   0
vehicletype           5095
registration_year        0
gearbox               2680
powerps                  0
model                 2758
odometer                 0
registration_month       0
fueltype              4482
brand                    0
unrepaired_damage     9829
ad_created               0
nrofpictures             0
postalcode               0
lastseen                 0
dtype: int64


In [9]:
# check columns price and odometer

# price column
print(autos["price"].dtype)
print(autos["price"].unique())

# odometer column
print(autos["odometer"].dtype)
print(autos["odometer"].unique())

object
['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']
object
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [10]:
# clean columns by:
# - remove non-numeric character
# - convert column to a numeric type
# - rename column to odometer_km

# price column cleanup step
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
print(autos["price"].unique())


[ 5000  8500  8990 ...   385 22200 16995]


In [11]:
# odometer column cleanup step
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
print(autos["odometer"].unique())

autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)
odometer_desc = autos["odometer_km"].describe()
print(odometer_desc)


[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [12]:
# print column names to check the renamed odometer column name
print(autos.columns)

Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer_km', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')


# 4) Task: Exploring the Odometer and Price Columns

In [13]:
# Check number of price unique values

print("unique values")
print(autos["price"].unique())

# describe column
print("")
print(autos["price"].describe())

# unique counts-head
print("\n")
print("value counts")
print("\n")
print("all counts")
print(autos["price"].value_counts())
print("\n")
print("descending order")
print(autos["price"].value_counts().head().sort_index(ascending=True))
print("\n")
print("ascending order")
print(autos["price"].value_counts().head().sort_index(ascending=False))
print("\n")
print("max price")
print(autos["price"].max())
print("\n")
print("min price")
print(autos["price"].min())

unique values
[ 5000  8500  8990 ...   385 22200 16995]

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


value counts


all counts
0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64


descending order
0       1421
500      781
1000     639
1500     734
2500     643
Name: price, dtype: int64


ascending order
2500     643
1500     734
1000     639
500      781
0       1421
Name: price, dtype: int64


max price
99999999


min price
0


In [14]:
# outliers

#print("unique values")
#print(autos["registration_year"].unique())

def display_unique(col_name):
    print(autos[col_name].unique())
    
for c in autos.columns:
    print("\n")
    print(c)
    display_unique(c)
    




datecrawled
['2016-03-26 17:47:46' '2016-04-04 13:38:56' '2016-03-26 18:57:24' ...
 '2016-03-28 10:50:25' '2016-03-08 19:25:42' '2016-03-14 00:42:12']


name
['Peugeot_807_160_NAVTECH_ON_BOARD'
 'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik' 'Volkswagen_Golf_1.6_United'
 ... 'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon'
 'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!'
 'Fiat_500_C_1.2_Dualogic_Lounge']


seller
['privat' 'gewerblich']


offertype
['Angebot' 'Gesuch']


price
[ 5000  8500  8990 ...   385 22200 16995]


abtest
['control' 'test']


vehicletype
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


registration_year
[2004 1997 2009 2007 2003 2006 1995 1998 2000 2017 2010 1999 1982 1990
 2015 2014 1996 1992 2005 2002 2012 2011 2008 1985 2016 1994 1986 2001
 2018 2013 1972 1993 1988 1989 1967 1973 1956 1976 4500 1987 1991 1983
 1960 1969 1950 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966
 1979 1981 1970 1974 1910 1975 5000

## Cleanup outliers for car prices.

In [15]:
# df[df["col"].between(x,y)]

print("\n")
print("max price")
print(autos["price"].max())
print("\n")
print("min price")
print(autos["price"].min())
print("\n")
print("unique")
print(autos["price"].unique())
print("\n")
print("null count")
print(autos["price"].isnull().sum())
print("\n")
print(autos["price"].value_counts(dropna=False))
# value counts shows that there are about 1421 rows that do not have a price

#check outliers before removing
print(autos[autos["price"].between(1,99999999)])




max price
99999999


min price
0


unique
[ 5000  8500  8990 ...   385 22200 16995]


null count
0


0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64
               datecrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone

# 5) Exploring the date columns


1. Calculate the distribution of the 3 date columns - date_crawled, ad_created, last_seen
2. Understand the distribution of the registration_year

In [16]:
# describe fields first

# display data to understand distribution w/o any extra processing
print(autos.describe())

# display datecrawled
print(autos['datecrawled'].str[:10])

# display ad_created
print(autos['ad_created'].str[:10])

# display ad_created
print(autos['lastseen'].str[:10])

# display all 3 dates
print(autos[['datecrawled','ad_created','lastseen']][0:5])

# display current values of the 3 date columns
#print(autos[['datec_rawled'].str[:10],['ad_created'].str[:10],['last_seen'].str[:10]][0:5])

              price  registration_year       powerps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nrofpictures    postalcode  
count        50000.000000       50000.0  50000.000000  
mean             5.723360           0.0  50813.627300  
std              3.711984           0.0  25779.747957  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  30451.000000  
50%              6.00000

In [17]:
# check distribution for datecrawled column

def display_unique(col_name):
    #print(autos[col_name].str[:10].value_counts(normalize=True, dropna=False).head().sort_index(ascending=True))
    dc = autos[col_name].str[:10].value_counts(normalize=True, dropna=False).head().sort_index(ascending=True)
    print(dc)
    
#datecrawled    
print("\n")
print("date_crawled distribution")
display_unique('datecrawled')





date_crawled distribution
2016-03-12    0.03678
2016-03-14    0.03662
2016-03-20    0.03782
2016-03-21    0.03752
2016-04-03    0.03868
Name: datecrawled, dtype: float64


###  Observation: 2016-04-03 for date_crawled column has a higher count than the other dates.

In [18]:
# check distribution for ad_created column


def display_unique(col_name):
    #print(autos[col_name].str[:10].value_counts(normalize=True, dropna=False).head().sort_index(ascending=True))
    ac = autos[col_name].str[:10].value_counts(normalize=True, dropna=False).head().sort_index(ascending=True)
    print(ac)

print("\n")
print("ad_created distribution")
display_unique('ad_created')



ad_created distribution
2016-03-12    0.03662
2016-03-20    0.03786
2016-03-21    0.03772
2016-04-03    0.03892
2016-04-04    0.03688
Name: ad_created, dtype: float64


### Observation: 2016-04-03 for ad_created column has a higher count than the other dates. Therefore, most ads were created on 2016-04-03.

In [19]:
# check distribution for lastseen column


def display_unique(col_name):
    #print(autos[col_name].str[:10].value_counts(normalize=True, dropna=False).head().sort_index(ascending=True))
    ls = autos[col_name].str[:10].value_counts(normalize=True, dropna=False).head().sort_index(ascending=True)
    print(ls)
    
print("\n")
print("last_seen distribution")
display_unique('lastseen')



last_seen distribution
2016-03-17    0.02792
2016-04-03    0.02536
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: lastseen, dtype: float64


### Observation: Most of the ads were most seen on 2016-04-06.

### Check the distribution of the registration year.



In [20]:
# check distribution for registration year.

autos.describe()
print(autos["registration_year"].describe())
print("\n")
print("value counts for registration year")
print(autos["registration_year"].value_counts(dropna=False).sort_index(ascending=False))

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


value counts for registration year
9999    4
9996    1
9000    2
8888    1
6200    1
       ..
1800    2
1500    1
1111    1
1001    1
1000    1
Name: registration_year, Length: 97, dtype: int64


### Observation: There are about 50000 registration years but about 97 unique values.

### The year with the max count is year 9999. The min year with the least count is year 1000.

# 6) Task: Dealing with incorrect year data

1. Decide which and lowest acceptable values are for the registration year.
2. Remove the values outside those upper and lower boudns and calculate the distribution of the remaining values.


In [21]:
# Desccribe registration year

print("\n")
print("Describe registration year")
print(autos["registration_year"].describe())

# Display uniqe years
print("\n")
print('Unique Registration Years')
print(autos["registration_year"].unique())
#autos["registration_year"].unique().sort_values(ascending=False)
unique_years = autos["registration_year"].unique()


# Display count of cars < 1900

print("\n")
print("Count of cars before the year 1900")
cars_before_1900 = autos["registration_year"] < 1900
print(cars_before_1900.value_counts())

# Display count of cars > 2016
print("\n")
print("Count of cars after the year 2016")
cars_afer_2016 = autos["registration_year"] > 2016
print(cars_afer_2016.value_counts())

# Display count of cars between 1900 - 2016
print("\n")
print("Count of cars between 1900 to 2016")
car_between_1900_2016 = autos["registration_year"].between(1900, 2016).value_counts()
print(car_between_1900_2016)


#Display min and max registration years
print("\n")
print("max registration year")
print(autos["registration_year"].max())
print("\n")
print("min registration_year")
print(autos["registration_year"].min())
print("\n")



Describe registration year
count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


Unique Registration Years
[2004 1997 2009 2007 2003 2006 1995 1998 2000 2017 2010 1999 1982 1990
 2015 2014 1996 1992 2005 2002 2012 2011 2008 1985 2016 1994 1986 2001
 2018 2013 1972 1993 1988 1989 1967 1973 1956 1976 4500 1987 1991 1983
 1960 1969 1950 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966
 1979 1981 1970 1974 1910 1975 5000 4100 2019 1959 9996 9999 6200 1964
 1958 1800 1948 1931 1943 9000 1941 1962 1927 1937 1929 1000 1957 1952
 1111 1955 1939 8888 1954 1938 2800 5911 1500 1953 1951 4800 1001]


Count of cars before the year 1900
False    49994
True         6
Name: registration_year, dtype: int64


Count of cars after the year 2016
False    48034
True      1966
Name: registration_year, dtype: int64


Count of cars between 1

### Observation: Valid years are from 1900 to 2016. Years over 2016 based on the data provided is not consired accurate.

### After deciding that the valid years are from 1900 to 2016, the other years will be removed from the count. The distribution of the valid years are shown belown.

In [22]:
# Display registration years for all

# Display count of cars < 1900

print("\n")
print("Count of cars before the year 1900")
cars_before_1900 = autos["registration_year"] < 1900
print(cars_before_1900.value_counts())

# Display count of cars > 2016
print("\n")
print("Count of cars after the year 2016")
cars_afer_2016 = autos["registration_year"] > 2016
print(cars_afer_2016.value_counts())

# Display count of cars > 2021
print("\n")
print("Count of cars between 1900 and 2016")
cars_range = autos["registration_year"].between(1900, 2016)
print(cars_range.value_counts())



Count of cars before the year 1900
False    49994
True         6
Name: registration_year, dtype: int64


Count of cars after the year 2016
False    48034
True      1966
Name: registration_year, dtype: int64


Count of cars between 1900 and 2016
True     48028
False     1972
Name: registration_year, dtype: int64


### Observation: There are about 48,028 registration years that is between 1900 and 2016.

### There are about 1,996 years after 2016 and 6 years before 1900.

### There are a total of 50,000 registration years.

### Adding all the counts - 48,028 + 1966+ 6 - sums to 50,000.

### Task below is to remove the upper and lower bounds to display the distribution of the selected values between 1900 and 2016.

In [23]:
# Remove outside values and display distribution

print("\n")
print(autos["registration_year"].between(1900,2016).value_counts(normalize=True))



True     0.96056
False    0.03944
Name: registration_year, dtype: float64


### Result from the distribution shows that the frequency of years between 1900 ad 2016 is .96056 out of the 50,000 total count for all registration years.


# 7) Task: Exploring Price by Brand

### Goal is to do the following process:
- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each: - Subset the dataframe by the unique values - Calculate the mean of whichever column we're insterested in - Assign the val/mean to the dict as k/v

### Task: Explore the unique vales in brand column and aggregate by:

- Select top 20 or select those that have a certain percentage value of the total values (e.g. > 5%)

In [24]:
#display unique brand
print("\n")
print("Display the brand unique values")
unique_brand = autos["brand"].unique()
print(unique_brand)

print("\n")
print("Describe brand")
print(autos["brand"].describe())

print("\n")
print("Describe price")
print(autos["price"].describe())

print("\n")
print("Display top 20 of the unique brand with counts in descending order")
top20 = (autos["brand"].value_counts(dropna=False).head(20).sort_values(ascending=False))
print(top20)





Display the brand unique values
['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


Describe brand
count          50000
unique            40
top       volkswagen
freq           10687
Name: brand, dtype: object


Describe price
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


Display top 20 of the unique brand with counts in descending order
volkswagen        10687
opel               5461
bmw                5429
mercedes_benz      4734
audi               4283
ford               3479
renault            2404


### The results above shows how the top 20 vehicle brands are displayed from the autos dataframe.

-Using the value counts displays the counts for each brand returned

-Dropna=False to exclude values of NA if applicable.

-Values are sorted in descending order, displaying the brand with the highest count at the top.

-The list shows that - Volkswagen is in the top 20 list for the vehicle brands.

In [25]:
#to iterate over the 20 brands
#loop through the brands and assign to the dictionary

#get top 20 values
top20 = (autos["brand"].value_counts(dropna=False).head(20).sort_values(ascending=False))

#create new dataframe for the top 20
unique_brand = pd.DataFrame(top20)
unique_brand = unique_brand.reset_index()
# change column names
unique_brand.columns = ['top_brand', 'mean price'] 


#create dictionary    
mean_price_by_top_20_branch = {}

#loop through the top 20 brands and slice the dataframe to get the brand name
for ub in unique_brand.iloc[:,0]: 
  #select brand from the autos data frame
  selected_rows = autos[autos["brand"] == ub]
  # calculate mean of price per brand
  mean = selected_rows["price"].mean() 
  # slice the row and assign brand name
  first_row = selected_rows.iloc[0]
  brand_name = first_row.loc["brand"]
  #assign values into the dictionary
  mean_price_by_top_20_branch[ub] = mean
  

print("\n")
print("Display dictionary for Top 20 Brands and mean value for price")
print("\n")
#print(avg_price_by_top_20_branch)
for key, value in mean_price_by_top_20_branch.items():
    print(key, ' : ', value)


    
#validate one brand
print("\n")
print("Validate one example for volkswagen")
brand = autos[autos["brand"] == "volkswagen"]
mean = brand["price"].mean() 
print(mean)





Display dictionary for Top 20 Brands and mean value for price


volkswagen  :  6384.167399644428
opel  :  5106.092657022524
bmw  :  8252.918953766808
mercedes_benz  :  29511.955428812842
audi  :  8965.560354891431
ford  :  7105.662546708824
renault  :  2351.301996672213
peugeot  :  3010.8688186813188
fiat  :  12134.20642201835
seat  :  4219.431455897981
skoda  :  6305.044529262086
mazda  :  3962.542932628798
nissan  :  4588.879310344828
smart  :  3482.971469329529
citroen  :  42657.46362339515
toyota  :  5097.941653160454
sonstige_autos  :  38300.84065934066
hyundai  :  5316.754098360656
volvo  :  31689.908096280087
mini  :  10392.393867924528


Validate one example for volkswagen
6384.167399644428


### The list above shows the mean price for the top 20 vehicle brands. Although the Volkswagen is on the top list for the top 20, its mean price is not the lowest.

### To iterate over the top 20 list, I created a different dataframe for the top 20 to use the brand names within the loop. The loop only computes the mean price just for the specific brands.




# 8) Task: Storing Aggregate Data in a Dataframe

### Goals for this task are:
- Use a loop method to calculate the mileage and mean price for each of the top brands
- Convert both dictionaries to series objects, using the series constructor
- Create a dataframe from the first series object using the dataframe constructor
- Assign the other series as new column in this dataframe
- Print the dataframe and write the analysis for the aggregated data

In [26]:
# describe autos df

autos.describe()

price  registration_year       powerps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nrofpictures    postalcode  
count        50000.000000       50000.0  50000.000000  
mean             5.723360           0.0  50813.627300  
std              3.711984           0.0  25779.747957  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  30451.000000  
50%              6.000000           0.0  49577.000000  
75%              9.000000           0.0  71540.000000  
max             12.000000           0.0  99998.000000

In [27]:
#previous dictionary for the brands' mean price

print("\n")
print("Display dictionary for Top 20 Brands and mean value for price")
print("\n")
#print(avg_price_by_top_20_branch)
for key, value in mean_price_by_top_20_branch.items():
    print(key, ' : ', value)





Display dictionary for Top 20 Brands and mean value for price


volkswagen  :  6384.167399644428
opel  :  5106.092657022524
bmw  :  8252.918953766808
mercedes_benz  :  29511.955428812842
audi  :  8965.560354891431
ford  :  7105.662546708824
renault  :  2351.301996672213
peugeot  :  3010.8688186813188
fiat  :  12134.20642201835
seat  :  4219.431455897981
skoda  :  6305.044529262086
mazda  :  3962.542932628798
nissan  :  4588.879310344828
smart  :  3482.971469329529
citroen  :  42657.46362339515
toyota  :  5097.941653160454
sonstige_autos  :  38300.84065934066
hyundai  :  5316.754098360656
volvo  :  31689.908096280087
mini  :  10392.393867924528


In [28]:
#create brand price dataframe and display value
print("\n")
print("Display new dataframe for brand mean price")
bmp_series = pd.Series(mean_price_by_top_20_branch)
dfmp = pd.DataFrame(bmp_series, columns=['mean_price'])
print(dfmp)




Display new dataframe for brand mean price
                  mean_price
volkswagen       6384.167400
opel             5106.092657
bmw              8252.918954
mercedes_benz   29511.955429
audi             8965.560355
ford             7105.662547
renault          2351.301997
peugeot          3010.868819
fiat            12134.206422
seat             4219.431456
skoda            6305.044529
mazda            3962.542933
nissan           4588.879310
smart            3482.971469
citroen         42657.463623
toyota           5097.941653
sonstige_autos  38300.840659
hyundai          5316.754098
volvo           31689.908096
mini            10392.393868


In [29]:
#get top 20 values
top20 = (autos["brand"].value_counts(dropna=False).head(20).sort_values(ascending=False))

#create new dataframe for the top 20
unique_mileage = pd.DataFrame(top20)
unique_mileage = unique_mileage.reset_index()
# change column names
unique_mileage.columns = ['top_brand', 'mean mileage'] 


#create dictionary    
mean_mileage_by_top_20_branch = {}

#loop through the top 20 brands and slice the dataframe to get the brand name
for ub in unique_brand.iloc[:,0]: 
  #select brand from the autos data frame
  selected_rows = autos[autos["brand"] == ub]
  # calculate mean of mileage per brand
  mean = selected_rows["odometer_km"].mean() 
  # slice the row and assign brand name
  first_row = selected_rows.iloc[0]
  brand_name = first_row.loc["brand"]
  #assign values into the dictionary
  mean_mileage_by_top_20_branch[ub] = mean
  

print("\n")
print("Display dictionary for Top 20 Brands and mean value for mileage (odometer_km)")
print("\n")
for key, value in mean_mileage_by_top_20_branch.items():
    print(key, ' : ', value)



Display dictionary for Top 20 Brands and mean value for mileage (odometer_km)


volkswagen  :  128955.27276129878
opel  :  129298.66324848929
bmw  :  132521.64302818198
mercedes_benz  :  130886.14279678918
audi  :  129643.9411627364
ford  :  124131.93446392642
renault  :  128223.79367720465
peugeot  :  127352.33516483517
fiat  :  117037.4617737003
seat  :  122061.63655685441
skoda  :  110947.83715012722
mazda  :  125132.10039630119
nissan  :  118978.7798408488
smart  :  100756.06276747503
citroen  :  119764.62196861627
toyota  :  115988.65478119935
sonstige_autos  :  87188.6446886447
hyundai  :  106782.7868852459
volvo  :  138632.3851203501
mini  :  89375.0


In [30]:
#previous dictionary for the brands' mean mileage

print("\n")
print("Display dictionary for Top 20 Brands and mean value for mileage")
print("\n")

for key, value in mean_mileage_by_top_20_branch.items():
    print(key, ' : ', value)



Display dictionary for Top 20 Brands and mean value for mileage


volkswagen  :  128955.27276129878
opel  :  129298.66324848929
bmw  :  132521.64302818198
mercedes_benz  :  130886.14279678918
audi  :  129643.9411627364
ford  :  124131.93446392642
renault  :  128223.79367720465
peugeot  :  127352.33516483517
fiat  :  117037.4617737003
seat  :  122061.63655685441
skoda  :  110947.83715012722
mazda  :  125132.10039630119
nissan  :  118978.7798408488
smart  :  100756.06276747503
citroen  :  119764.62196861627
toyota  :  115988.65478119935
sonstige_autos  :  87188.6446886447
hyundai  :  106782.7868852459
volvo  :  138632.3851203501
mini  :  89375.0


In [31]:
#create brand price dataframe and display value
print("\n")
print("Display new dataframe for brand mean mileage")
bm_series = pd.Series(mean_mileage_by_top_20_branch)
dfmm = pd.DataFrame(bm_series, columns=['mean_mileage'])
print(dfmm)




Display new dataframe for brand mean mileage
                 mean_mileage
volkswagen      128955.272761
opel            129298.663248
bmw             132521.643028
mercedes_benz   130886.142797
audi            129643.941163
ford            124131.934464
renault         128223.793677
peugeot         127352.335165
fiat            117037.461774
seat            122061.636557
skoda           110947.837150
mazda           125132.100396
nissan          118978.779841
smart           100756.062767
citroen         119764.621969
toyota          115988.654781
sonstige_autos   87188.644689
hyundai         106782.786885
volvo           138632.385120
mini             89375.000000


### Analysis of the results for No. 8 Tasks


The analsysis that I have done is based on the top 20 brands.

1. Mean Mileage

The smallest mean mileage is for the Sonstige Autos with a a value of 87188.644689.

The highest mean mileage is for Volvo with a value of 138632.385120.

2. Mean Price

The smallest mean price is for the Renault with a value of 2351.301997.

The highest mean price is Citroen with a value 42657.463623.


# 9) Task: Last Step in this project


## Goals for this task are:

1. Data Cleaning Steps
2. Analysis Steps


### 1) Data Cleaning Steps

The Cleaning Steps demonstrates the following:

1. Convert German words, translate and map to their English counterparts.
2. Convert the dates to be uniform numeric data. From "2016-03-21" to an integer format of 20160321.
3. Extract particular keywords in the name column that can be extracted as a new column.


In [32]:
# display columns from the data frame to review
# note that other columns have been cleaned in the previous steps

print(autos.columns)
print("\n")
# describe autos dataframe
autos.describe()


Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'registration_year', 'gearbox', 'powerps', 'model',
       'odometer_km', 'registration_month', 'fueltype', 'brand',
       'unrepaired_damage', 'ad_created', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')




price  registration_year       powerps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nrofpictures    postalcode  
count        50000.000000       50000.0  50000.000000  
mean             5.723360           0.0  50813.627300  
std              3.711984           0.0  25779.747957  
min              0.000000           0.0   1067.000000  
25%              3.000000           0.0  30451.000000  
50%              6.000000           0.0  49577.000000  
75%              9.000000           0.0  71540.000000  
max             12.000000           0.0  99998.000000

In [33]:
# check unique values for each column for review and identify
# columns with german words
# reusing function in previous step

#function display_unique values for each column
def display_unique(col_name):
    print(autos[col_name].unique())
    
for c in autos.columns:
    print("\n")
    print(c)
    display_unique(c)



datecrawled
['2016-03-26 17:47:46' '2016-04-04 13:38:56' '2016-03-26 18:57:24' ...
 '2016-03-28 10:50:25' '2016-03-08 19:25:42' '2016-03-14 00:42:12']


name
['Peugeot_807_160_NAVTECH_ON_BOARD'
 'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik' 'Volkswagen_Golf_1.6_United'
 ... 'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon'
 'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!'
 'Fiat_500_C_1.2_Dualogic_Lounge']


seller
['privat' 'gewerblich']


offertype
['Angebot' 'Gesuch']


price
[ 5000  8500  8990 ...   385 22200 16995]


abtest
['control' 'test']


vehicletype
['bus' 'limousine' 'kleinwagen' 'kombi' nan 'coupe' 'suv' 'cabrio'
 'andere']


registration_year
[2004 1997 2009 2007 2003 2006 1995 1998 2000 2017 2010 1999 1982 1990
 2015 2014 1996 1992 2005 2002 2012 2011 2008 1985 2016 1994 1986 2001
 2018 2013 1972 1993 1988 1989 1967 1973 1956 1976 4500 1987 1991 1983
 1960 1969 1950 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966
 1979 1981 1970 1974 1910 1975 5000

### Based on the results displayed, the "gearbox" columns will be used as an example for the step to convert the German word to an English word.

#### Tasks

1. Convert 'manuell' to 'manual'
2. Convert 'automatik' to 'automatic'



In [34]:
# Use replace function to change values for the "gearbox" column

# gearbox column cleanup step
autos["gearbox"] = autos["gearbox"].str.replace("manuell","manual").str.replace("automatik","automatic")
print(autos["gearbox"].unique())

['manual' 'automatic' nan]


### The German words have been converted into its matching English counterparts.

### On to changing the date formats for the date columns.

In [35]:
# The columns to reformat are: datecrawled, ad_created, lastseen
# note that lastseen column has time values


#display information about autos

print(autos.info())
print("\n")
print(autos.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   datecrawled         50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offertype           50000 non-null  object
 4   price               50000 non-null  int64 
 5   abtest              50000 non-null  object
 6   vehicletype         44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   powerps             50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fueltype            45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

In [36]:
# clean columns by:
# - remove non-numeric character
# - convert column to an integer


# clean-up date steps
def clean_date(col_name):
    autos[col_name] = autos[col_name].str[:10].str.replace("-","").astype(int)


# date column cleanup step
print("\n")
#print("Display datecrawled values")
#print(autos["datecrawled"].str[:10])
#clean datecrawled
clean_date("datecrawled")
clean_date("ad_created")
clean_date("lastseen")

print("\n")
print("Display sample of converted dates into new format")
#utos["datecrawled"] = autos["datecrawled"].str[:10].str.replace("-","").astype(int)
#print(autos["datecrawled"].str[:10].str.replace("-","").astype(int))
print("\n")
print(autos[['datecrawled','ad_created','lastseen']][0:10])






Display sample of converted dates into new format


   datecrawled  ad_created  lastseen
0     20160326    20160326  20160406
1     20160404    20160404  20160406
2     20160326    20160326  20160406
3     20160312    20160312  20160315
4     20160401    20160401  20160401
5     20160321    20160321  20160406
6     20160320    20160320  20160323
7     20160316    20160316  20160407
8     20160322    20160322  20160326
9     20160316    20160316  20160406


### On to finding particular keywords in the name column

In [37]:
# check for keywords in the name column

#function display_unique values for each column
def display_unique(col_name):
    print("Display unique values for " + col_name)
    print("\n")
    print(autos[col_name].unique())
    
display_unique("name")


Display unique values for name


['Peugeot_807_160_NAVTECH_ON_BOARD'
 'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik' 'Volkswagen_Golf_1.6_United'
 ... 'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon'
 'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!'
 'Fiat_500_C_1.2_Dualogic_Lounge']


In [38]:
# add new column by using unique values from the contents displayed above
# new column to add will be named as "name_keyword"
# Example. "Peugeot_807_160_NAVTECH_ON_BOARD" will have the keyword of "Peugeot"

# assign cleaned columns names to a new array
# function created to clean columns

#print(autos["name"].head())

keywords = ['Peugeot', 'BMW', 'Audi', 'Volkswagen', 'Ford', 'Fiat', 'Opel', 'Astra', 'Smart']

autos['name_keyword'] = autos['name'].str.findall('|'.join(keywords)).apply(set).str.join(', ')

print("\n")
print("Display the top 5 new keywords")
print(autos["name_keyword"].head())
print(autos[['name','name_keyword']][0:5])





Display the top 5 new keywords
0       Peugeot
1           BMW
2    Volkswagen
3         Smart
4          Ford
Name: name_keyword, dtype: object
                                                name name_keyword
0                   Peugeot_807_160_NAVTECH_ON_BOARD      Peugeot
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik          BMW
2                         Volkswagen_Golf_1.6_United   Volkswagen
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...        Smart
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...         Ford


### 2)  Analysis Steps
### Find the most common/brand models

In [39]:
#Display the brand and model columns only for the top 20 as a sample

print("\n")
print("Top combination of brand and model")
top20 = (autos["brand"].value_counts(dropna=False).head(20).sort_values(ascending=False))


#create dictionary    
common_brand_model = {}

#create new dataframe for the top 20
unique_brand = pd.DataFrame(top20)
unique_brand = unique_brand.reset_index()
# change column names
unique_brand.columns = ['brand', 'model'] 

#loop through the top 20 brands and slice the dataframe to get the brand name
for ub in unique_brand.iloc[:,0]: 
  #select brand from the autos data frame
  selected_rows = autos[autos["brand"] == ub]
  # slice the row and assign brand name
  first_row = selected_rows.iloc[0]
  model_name = first_row.loc["model"]
  #assign values into the dictionary
  common_brand_model[ub] = model_name
  

print("Display dictionary for Top 20 Brands and mean value for price")
#print(avg_price_by_top_20_branch)
#for key, value in common_brand_model.items():
#    print(key, '  :  ', value)
    
    
#create brand price dataframe and display value
print("\n")
print("Display new dataframe for brand and model")
bm_series = pd.Series(common_brand_model)
dfbm = pd.DataFrame(bm_series, columns=['model'])
print(dfbm)

#display brand mean price km dataframe and display value
print("\n")
print("Display new dataframe for brand mean price")
bmp_series = pd.Series(mean_price_by_top_20_branch)
dfmp = pd.DataFrame(bmp_series, columns=['mean_price'])
print(dfmp)

#display mean mileage dataframe and display value
print("\n")
print("Display new dataframe for brand mean mileage")
bm_series = pd.Series(mean_mileage_by_top_20_branch)
dfmm = pd.DataFrame(bm_series, columns=['mean_mileage'])
print(dfmm)






Top combination of brand and model
Display dictionary for Top 20 Brands and mean value for price


Display new dataframe for brand and model
                  model
volkswagen         golf
opel             vectra
bmw                 7er
mercedes_benz       NaN
audi                 a3
ford              focus
renault          megane
peugeot          andere
fiat              punto
seat              arosa
skoda           octavia
mazda            andere
nissan          primera
smart            fortwo
citroen          andere
toyota           andere
sonstige_autos      NaN
hyundai         i_reihe
volvo            andere
mini             cooper


Display new dataframe for brand mean price
                  mean_price
volkswagen       6384.167400
opel             5106.092657
bmw              8252.918954
mercedes_benz   29511.955429
audi             8965.560355
ford             7105.662547
renault          2351.301997
peugeot          3010.868819
fiat            12134.206422
seat             4

### Analysis from above:

The sonstige_autos had the lowest mileage and the price is high compared to other cars with a higher mileage like citroen.

There are no descriptive values for the unrepaired_damage column.

Renault's mean price is lower than the others and it has a mileage but not the highest of the top 20.